src/shm/pool/slot.rs

```diff
@@ -0,0 +1,412 @@
+//! A pool implementation based on buffer slots
+
+use std::io;
+use std::sync::{
+    atomic::{AtomicU8, AtomicUsize, Ordering},
+    Arc, Mutex, Weak,
+};
+
+use wayland_client::{
+    backend::InvalidId,
+    protocol::{wl_buffer, wl_shm, wl_surface},
+    Connection, Proxy,
+};
+
+use super::raw::RawPool;
+
+#[derive(Debug, thiserror::Error)]
+pub enum CreateBufferError {
+    /// Protocol error.
+    #[error(transparent)]
+    Protocol(#[from] InvalidId),
+
+    /// Slot creation error.
+    #[error(transparent)]
+    Io(#[from] io::Error),
+
+    /// Pool mismatch.
+    #[error("Incorrect pool for slot")]
+    PoolMismatch,
+
+    /// Slot size mismatch
+    #[error("Buffer too large for slot")]
+    SlotTooSmall,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ActivateSlotError {
+    /// Buffer was already active
+    #[error("Buffer was already active")]
+    AlreadyActive,
+}
+
+#[derive(Debug)]
+pub struct SlotPool {
+    pub(crate) inner: RawPool,
+    free_list: Arc<Mutex<Vec<(usize, usize)>>>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Slot {
+    inner: Arc<SlotInner>,
+}
+
+#[derive(Debug)]
+struct SlotInner {
+    free_list: Weak<Mutex<Vec<(usize, usize)>>>,
+    offset: usize,
+    len: usize,
+    active_buffers: AtomicUsize,
+}
+
+/// Buffer is counted in active_buffers list; will return to INACTIVE on Release.
+const BUFFER_STATE_ACTIVE: u8 = 0;
+
+/// Buffer is not counted in active_buffers list, but also has not been destroyed.
+const BUFFER_STATE_INACTIVE: u8 = 1;
+
+/// Buffer is counted in active_buffers list; will move to DEAD on Release
+const BUFFER_STATE_DESTROY_ON_RELEASE: u8 = 2;
+
+/// Buffer has been destroyed
+const BUFFER_STATE_DEAD: u8 = 3;
+
+/// Value that is ORed on buffer release to transition to the next state
+const BUFFER_STATE_RELEASE_SET: u8 = 1;
+
+/// Value that is ORed on buffer destroy to transition to the next state
+const BUFFER_STATE_DESTROY_SET: u8 = 2;
+
+#[derive(Debug)]
+pub struct Buffer {
```

Comment:
It would be quite nice to be able to use a buffer to get the area to draw into vs getting the entire slot's content and calculating where to draw myself